### PR TITLE
feat: Laravel construct

### LINKED ISSUE #42: Laravel high-level construct
Now that we are going to have a working sample for Laravel  https://github.com/pahud/cdk-fargate-patterns/pull/41

Similar to [WordPress()](https://github.com/pahud/cdk-fargate-patterns/blob/ec9dc6ee8c7acf775b169a8a203d6d7275dd40a4/src/wordpress.ts#L72), wondering if we could have a high-level construct like ` new Laravel()` and everything would be all set?

The building experience should be like

```ts

new Laravel(scope, id, {
   spot: true,  // make it FARGATE_SPOT
   path,  // path to local Laravel code base
}
```

@dwchiang @clarencetw @neilkuan  wdyt?

### PR DESCRIPTION
**Pull request by pahud**
1. Introduce the `Laravel` high-level construct with samples.
2. Refactor `WordPress`.  Move the database class to the `common` directory.
3. Move `OrderService` and `CommonService` to `service/golang`

Fixes #42

### CODE REVIEW

**Comment by pahud** [comment]
@dwchiang @neilkuan  please take a look.

File: src/laravel.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        fileSystemId: filesystem.fileSystemId,
+      },
+    });
+
+    // fix me - tentatively mount to /efsmount
+    this.svc.service[0].taskDefinition.defaultContainer?.addMountPoints({
+      containerPath: '/efsmount',
+      readOnly: false,
+      sourceVolume: volumeName,
+    });
```
Comment (pahud) [inline src/laravel.ts:177]: Hi @dwchiang 

Do we need EFS mount for the Laravel environment? If not, I am happy to remove it.

File: src/common/db.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    };
+  }
+  // create a RDS for MySQL DB cluster
+  private _createRdsCluster(props: DatabaseProps): DatabaseCofig {
+    const dbCluster = new rds.DatabaseCluster(this, 'DBCluster', {
+      engine: props.clusterEngine ?? rds.DatabaseClusterEngine.auroraMysql({
+        version: rds.AuroraMysqlEngineVersion.VER_2_09_1,
+      }),
+      credentials: rds.Credentials.fromGeneratedSecret('admin'),
+      defaultDatabaseName: 'Laravel',
```
Comment (neilkuan) [inline src/common/db.ts:139]: I think here can not hard code database name ?!

File: src/common/db.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  private _createServerlessCluster(props: DatabaseProps): DatabaseCofig {
+    const dbCluster = new rds.ServerlessCluster(this, 'AuroraServerlessCluster', {
+      engine: rds.DatabaseClusterEngine.AURORA_MYSQL,
+      vpc: props.vpc,
+      vpcSubnets: props.databaseSubnets,
+      credentials: rds.Credentials.fromGeneratedSecret('admin'),
+      backupRetention: props.backupRetention ?? cdk.Duration.days(7),
+      removalPolicy: cdk.RemovalPolicy.DESTROY,
+      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.aurora-mysql5.7'),
+      defaultDatabaseName: 'Laravel',
```
Comment (neilkuan) [inline src/common/db.ts:168]: Here also.

**Review by neilkuan — CHANGES_REQUESTED** [review]

File: src/common/db.ts (hunk 2 above)
Comment (pahud) [reply src/common/db.ts:139]: yes we should allow user to set the default database name if any. I'll revise it tomorrow.

**Comment by pahud** [comment]
@neilkuan please check it out again.

File: src/laravel.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+          external: props.loadbalancer,
+          task,
+          healthCheck,
+        },
+      ],
+      route53Ops: { enableLoadBalancerAlias: false },
+    });
+
+    // EFS volume
+    const filesystem = new efs.FileSystem(this, 'FileSystem', {
```
Comment (neilkuan) [inline src/laravel.ts:167]: Does efs need to be kept after destroy stack?!

File: src/common/db.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      engine: props.instanceEngine ?? rds.DatabaseInstanceEngine.mysql({
+        version: rds.MysqlEngineVersion.VER_8_0_23,
+      }),
+      storageEncrypted: true,
+      backupRetention: props.backupRetention ?? cdk.Duration.days(7),
+      credentials: rds.Credentials.fromGeneratedSecret('admin'),
+      instanceType: props.instanceType ?? new ec2.InstanceType('r5.large'),
+      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.mysql8.0'),
+      removalPolicy: cdk.RemovalPolicy.DESTROY,
+    });
```
Comment (neilkuan) [inline src/common/db.ts:130]:
```suggestion
    const dbInstance = new rds.DatabaseInstance(this, 'DBInstance', {
      vpc: props.vpc,
      vpcSubnets: props.databaseSubnets,
      databaseName: props.defaultDatabaseName,
      engine: props.instanceEngine ?? rds.DatabaseInstanceEngine.mysql({
        version: rds.MysqlEngineVersion.VER_8_0_23,
      }),
      storageEncrypted: true,
      backupRetention: props.backupRetention ?? cdk.Duration.days(7),
      credentials: rds.Credentials.fromGeneratedSecret('admin'),
      instanceType: props.instanceType ?? new ec2.InstanceType('r5.large'),
      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.mysql8.0'),
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
```

Comment (neilkuan) [reply src/common/db.ts:130]: I think DatabaseInstance need this `databaseName: props.defaultDatabaseName` for single DB Instance.

**Review by neilkuan — CHANGES_REQUESTED** [review]

File: src/laravel.ts (hunk 4 above)
Comment (pahud) [reply src/laravel.ts:167]: Interesting. We probably should destroy by default while leave the option for user to opt out.

File: src/laravel.ts (hunk 1 above)
Comment (dwchiang) [reply src/laravel.ts:185]: I think EFS is a good one, but not mandatory. Let's put it as an optional on/off parameter?

- Some (or most) legacy PHP application which usually does not store files in object-base storage (e.g. S3), they need EFS for smooth migration. e.g. WordPress (https://aws.amazon.com/blogs/storage/optimizing-wordpress-performance-with-amazon-efs/)
- Modern PHP applications may start to consider with object-base storage solution, they may only use the temp folder for file uploading usage. All the other file level access should move to work with object-base storage. :)

File: src/laravel.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+      // fix me - tentatively mount to /efsmount
+      this.svc.service[0].taskDefinition.defaultContainer?.addMountPoints({
+        containerPath: '/efsmount',
+        readOnly: false,
+        sourceVolume: volumeName,
+      });
+
+      filesystem.connections.allowFrom(new ec2.Connections({ securityGroups: this.svc.service[0].connections.securityGroups }), ec2.Port.tcp(2049), 'allow Laravel to connect to efs filesystem');
+    }
```
Comment (pahud) [inline src/laravel.ts:191]: Now I made the efs optional.

If `props.efsFileSystem` is provided, we provision a efs filesystem for it. In this case, which mount point should we mount to? I tentatively mount to `/efsmount` now.

@dwchiang  wdyt?

Comment (pahud) [reply src/laravel.ts:191]: cc @neilkuan

File: src/integ.laravel.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+new Laravel(stack, 'LaravelNginxDemo', {
+  vpc,
+  auroraServerless: true,
+  spot: true,
+  enableExecuteCommand: true,
+  code: path.join(__dirname, '../services/laravel-nginx-php-fpm'),
+  loadbalancer: cert ? { port: 443, certificate: [cert] } : { port: 80 },
+  efsFileSystem: {
+    vpc,
+    removalPolicy: cdk.RemovalPolicy.DESTROY,
```
Comment (pahud) [inline src/integ.laravel.ts:31]: Hey @neilkuan how about this?

Comment (neilkuan) [reply src/integ.laravel.ts:31]: LGTM

**Review by neilkuan — APPROVED** [review]
LGTM

**Pull request merged**